internal/controller/dnsrecord_healthchecks.go

```diff
@@ -2,180 +2,161 @@ package controller
+		logger.V(1).Info(fmt.Sprintf("Deleting probe: %s", probe.Name))
+		if err := r.Delete(ctx, &probe); err != nil {
+			return err
+		}
 	}
+	return nil
+}
 
-	dnsRecord.Status.HealthCheck.Probes = []v1alpha1.HealthCheckStatusProbe{}
+func (r *DNSRecordReconciler) ensureProbe(ctx context.Context, generated *v1alpha1.DNSHealthCheckProbe, logger logr.Logger) error {
+	current := &v1alpha1.DNSHealthCheckProbe{}
 
-	for _, result := range results {
-		if result.Host == "" {
-			continue
-		}
-		status := true
-		if result.Result == provider.HealthCheckFailed {
-			status = false
-			allSynced = metav1.ConditionFalse
+	if err := r.Get(ctx, client.ObjectKeyFromObject(generated), current); err != nil {
+		if errors.IsNotFound(err) {
+			logger.V(1).Info(fmt.Sprintf("Creating probe: %s", generated.Name))
+			return r.Create(ctx, generated)
 		}
-
-		dnsRecord.Status.HealthCheck.Probes = append(dnsRecord.Status.HealthCheck.Probes, v1alpha1.HealthCheckStatusProbe{
-			ID:         result.ID,
-			IPAddress:  result.IPAddress,
-			Host:       result.Host,
-			Synced:     status,
-			Conditions: []metav1.Condition{result.Condition},
-		})
+		return err
 	}
 
-	probesCondition.ObservedGeneration = dnsRecord.Generation
-	probesCondition.Status = allSynced
-
-	if allSynced == metav1.ConditionTrue {
-		probesCondition.Message = fmt.Sprintf("all %v probes synced successfully", len(dnsRecord.Status.HealthCheck.Probes))
-		probesCondition.Reason = "AllProbesSynced"
-	} else {
-		probesCondition.Reason = "UnsyncedProbes"
-		probesCondition.Message = "some probes have not yet successfully synced to the DNS Provider"
-	}
+	desired := current.DeepCopy()
+	desired.Spec = generated.Spec
 
-	//probe condition changed? - update transition time
-	if !reflect.DeepEqual(previousCondition, probesCondition) {
-		probesCondition.LastTransitionTime = metav1.Now()
+	if !reflect.DeepEqual(current, desired) {
+		logger.V(1).Info(fmt.Sprintf("Updating probe: %s", desired.Name))
+		if err := r.Update(ctx, desired); err != nil {
+			return err
+		}
 	}
-
-	dnsRecord.Status.HealthCheck.Conditions = []metav1.Condition{*probesCondition}
-
+	logger.V(1).Info(fmt.Sprintf("No updates needed for probe: %s", desired.Name))
 	return nil
 }
 
-func getHealthChecksConfig(dnsRecord *v1alpha1.DNSRecord) *healthChecksConfig {
-	if dnsRecord.Spec.HealthCheck == nil || dnsRecord.DeletionTimestamp != nil {
-		return nil
-	}
-
-	port := int64(dnsRecord.Spec.HealthCheck.Port)
-	failureThreshold := int64(dnsRecord.Spec.HealthCheck.FailureThreshold)
-
-	return &healthChecksConfig{
-		Endpoint:         dnsRecord.Spec.HealthCheck.Path,
-		Port:             &port,
-		FailureThreshold: &failureThreshold,
-		Protocol:         (*provider.HealthCheckProtocol)(&dnsRecord.Spec.HealthCheck.Protocol),
+func buildDesiredProbes(dnsRecord *v1alpha1.DNSRecord, leafs *[]string, allowInsecureCerts bool) []*v1alpha1.DNSHealthCheckProbe {
+	var probes []*v1alpha1.DNSHealthCheckProbe
+
+	for _, leaf := range *leafs {
```

Comment:
should do a nil pointer check

Reply:
Hmm, at first glance I thought that it would never happen, but, in fact, it could if we have nil for the DNS record. Added a check